### PR TITLE
2022.6.2 dev rrnf

### PR DESCRIPTION
**Pull request by hlyi**
# What does this implement/fix?
This implement has three parts:

1.  remote_receiver_nf component: similar to remote_receiver, it stores time of leading/falling edges in a ring buffer. There are three noticeable differences:
  - it can handle very noisy signal: for example, the CMT2210 RF 433MHz receiver keeps toggling at the rate of a few kHz even  there is no signal. A lot of other cheap RF receivers have similar characteristics. With users defined timing of "space" of the sync signal together with other techniques, this new component could filter out 99+% of noise.
  - it can break up back-to-back repeating codes. For example, the idle time between two repeating codes of Govee water leak sensor is only 1.1ms signal, i.e. the timing of a single bit signal. This implementation could handle it well.
  - On esp32, it uses GPIO interrupt, similar to ESP8266, because hardware RMT in ESP32 doesn't meet the algorithm requirement
2. Govee protocol component. It is for decoding 48bit code used by Govee sensors (tested on water leak sensor) Thanks to  users who reserve-engineered protocol that I found in this thread (https://community.home-assistant.io/t/govee-wifi-water-sensor/213277/78) . The component also de-duplicated codes, which Govee sensors send 12-13 times. The limitation is only one Govee decoder is supported (due to static variables for tracking time). Combined with remote_receiver_nf, no single false code has been reported for 24 hours, and no missing detection either. This component is independent of remote_receiver_nf. it may be able to decode Govee protocol with existing remote_receiver if RF signal was clean, but it hasn't been tested.
3. govee2mqtt component. This component is to take decoded results from Govee protocol component, and send them to MQTT broker with HA MQTT Discovery support. It addresses another shortcoming of Govee sensor: it only sends "on" signal, but not "off" signal. e.g when water leak detected, it sends "leak" code every 5-6 seconds, when the leak condition removed, it just stops sending "leak" codes instead of sending a "dry" code. Similarly, for push button, it sends the code only when the button is released.   The component generates both "on" and (generated) "off" payloads to MQTT, so no additional work on HA side is required.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** 

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** https://github.com/esphome/esphome-docs/pull/2178

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [x] ESP8266

## Example entry for `config.yaml`:

```yaml
mqtt:
  broker: 1.1.1.1
  id: mqtt_client
  topic_prefix: govee2mqtt/gateway
globals:
- id: devices
  type: std::map<int, std::vector<std::string>>
  initial_value: '{{0x1122, {"H5040", "My Water Leak Sensor"}}}'
remote_receiver_nf:
- pin:
    number: 12
    inverted: false
  id: govee_rf_rx
  tolerance: 20
  filter: 100us
  idle: 10ms
  buffer_size: 1024
  sync_space_min: 8300us
  sync_space_max: 9500us
  early_check_thres: 100
  sync_space_is_high: false
  repeat_space_min: 1100us
  num_edge_min: 16
govee2mqtt:
- dev_list: devices
  mqtt_id: mqtt_client
  receiver_id: govee_rf_rx
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/dsmr/__init__.py [hunk 1]
```diff
@@ -79,7 +79,7 @@ async def to_code(config):
     cg.add(var.set_request_interval(config[CONF_REQUEST_INTERVAL].total_milliseconds))
     cg.add(var.set_receive_timeout(config[CONF_RECEIVE_TIMEOUT].total_milliseconds))
```
Comment (mmakaay) [inline esphome/components/dsmr/__init__.py]: Something fishy is going on here with a commit to the mqtt component.

**Comment by mmakaay** [comment]
The commit touches some unrelated components. That needs some cleanup.

On first glance, another thing that I notice is the govee2mqtt component. Why did you choose to make that into a component, instead of providing some sensor outputs that can be exposed through mqtt by means of simply adding `mqtt:` to the device yaml config?

File: esphome/components/dsmr/__init__.py (hunk 1 above)
Comment (hlyi) [reply esphome/components/dsmr/__init__.py]: I didn't touch this file, don't know why this one show up, my local merge didn't see change in this file either.

**Comment by hlyi** [comment]
> 

by "touches some unrelated components", did you mean `dsmr/__init__.py`? If so, I didn't touch the file. I don't know if because my development is based upon 2022.6.2 tag.

qovee2mqtt is to provide ease of use. 

In theory, with govee protocol component, the raw information is already available. User could use on_govee  for automation.  However, sensors simply broadcast event without any encryption, and the range of 433MHz radio is quite long, even cheap RF receiver could pick up neighbors' sensors easily.  Filtering irrelevant sensors is highly desirable. With qovee2mqtt component,  a simple configuration of allowed device list in globals section provides:

- the fix for the shortcoming of sensors that not sending "off" code
- auto generate config for listed sensors, so HA MQTT discovery can pick up sensors without any yaml manipulation. 
- filter irrelevant sensors

In summary, qovee2mqtt is not required for govee sensors to work with esphome, but nice to have for easy integration.

**Comment by martgras** [comment]
You are touching unlreated components because you are not basesing the PR on dev. As a result there are changes files from other PRs in yours (sgp4x for example)

You should rebase your PR to the latest dev tree

**Comment by mmakaay** [comment]
> In summary, qovee2mqtt is not required for govee sensors to work with esphome, but nice to have for easy integration.

Sorry, but I still don't understand why standard sensors + mqtt component does not handle what you want to accomplish. If you drop `ota:` and use `mqtt:` instead, then sensors ought to popup in Home Assistant though the auto-discovery. 
For example the light component is perfectly usable over MQTT. There's not a `light2mqtt` component though. This goes for every other component in the system.

**Comment by hlyi** [comment]
> 
Just re-based the code. Please let me know if you still saw an issue.

**Comment by hlyi** [comment]
> Sorry, but I still don't understand why standard sensors + mqtt component does not handle what you want to accomplish. If you drop `ota:` and use `mqtt:` instead, then sensors ought to popup in Home Assistant though the auto-discovery. For example the light component is perfectly usable over MQTT. There's not a `light2mqtt` component though. This goes for every other component in the system.

First, recap what I tried to do here:
1. remote_receiver_nf is for addressing physical layer issue with hardware
2. govee protocol is for decoding data + squash repeating code (IMHO, it is spill-over from physical layer)
3. govee2mqtt is to address 
  - deficiency of network layer (not sending "off" code)
  - easy integration: each sensor has multiple entities. I would like to have HA seen the device in the same way, instead of multiple devices. maybe I missed the setting, how to do in esphome mqtt?

If needed, I could remove govee2mqtt from this pull request, and create a separate request. 1 and 2 are essential.

**Comment by mmakaay** [comment]
> * maybe I missed the setting, how to do in esphome mqtt?

https://esphome.io/components/mqtt.html
"All devices defined through ESPHome should show up automatically in the entities section of Home Assistant."

**Comment by hlyi** [comment]
> 



> https://esphome.io/components/mqtt.html
> "All devices defined through ESPHome should show up automatically in the entities section of Home Assistant."

Sorry, I didn't make my statement clear earlier. Maybe I missed something, I saw all devices from one esphome becomes entities in HA instead of devices with their own entities.  For example, if a esphome device controls 10 govee water leak sensors, there would be 30 entities + a few from esphome itself under single device in HA. It is not very convenient to re-assign 30 entities to different locations(rooms). If there were shown as 10 devices in HA and each device has 3 entities (this configuration matches real device behavior), it would be easier just re-assign 10 devices to different locations, basically, reduce manual work by factor of 3. qovee2mqtt does this way.

**Comment by ssieb** [comment]
This won't be accepted like that.  You can't override the framework, especially when it's specific to a transport.  You can keep this as an external component.
Also, sending the "off" code has nothing to do with mqtt.  You can do that generally.  That's how the remote receiver binary sensors work.

**Comment by mmakaay** [comment]
Also wondering what you'd get when using this specific mqtt component and also adding ESPHome's  mqtt component to the config.

**Comment by hlyi** [comment]
I removed qovee2mqtt from pull branch.

**Comment by mmakaay** [comment]
> > 
> 
> Just re-based the code. Please let me know if you still saw an issue.

Well, if you go to the "Files Changed" tab, you can easily see for yourself that the PR still touches a lot of other components. If you rebased your local working branch on dev, then maybe you haven't pushed the rebase outcome to github? You most likely will have to do a force push, because your fork on GitHub and your local repo clone have diverged.

**Comment by hlyi** [comment]
Thanks for the tip, @mmakaay. Just re-submitted the changes.

**Comment by hlyi** [comment]
How to address "needs-tests" tag?

**Comment by martgras** [comment]
You have to add your test config to one of the yaml files in the tests folder.

**Comment by hlyi** [comment]
I saw 5 test yaml files directly under tests folder. Which one  should I use?

**Comment by mmakaay** [comment]
Just pick one where it doesn't lead to conflicts of some sort.
The first one is fine if that works for you.

**Comment by hlyi** [comment]
Thanks @mmakaay. I added the test. I also created the document (https://github.com/esphome/esphome-docs/pull/2178). But tag still shows "needs-docs". How to resolve it?

**Comment by mmakaay** [comment]
That tag is handled by the reviewers.

**Comment by hlyi** [comment]
Got it. Anything I need to do at this point?

**Comment by mmakaay** [comment]
I think you got it all, so now wait for the reviews.

**Comment by mmakaay** [comment]
Oh, you didn't link the doc PR in the "Pull request in esphome-docs with documentation (if applicable):" section in the starting message. And there's also a checkbox to indicate that you took care of the documentation. Please update the starting message with that info.

**Comment by hlyi** [comment]
Thanks. Updated original pull request.
